### ISSUE TITLE
Spearman's rank correlation coefficient slow with nan values and nan policy = omit

### ISSUE THREAD

**Issue by jaksmid** [scipy/scipy#6654]
Hello,

I have two arrays of sizes around 8000 reals. The second array contains some nan values.
I want to calculate spearmans rank correlation coefficient and would like to ignore pairs containing nans.

According to the documentation 

>  nan_policy ‘omit’ performs the calculations ignoring nan values. 

So I set the nan_policy to omit. After ten minutes of computation I cancelled the computation.
I examined the underlying code and discovered that nan values are assigned the rank of zero.
Shouldnt they be skipped altogether?

When I ignored the nan values myself like this

```
from numpy import ma
x1 = ma.masked_invalid(array1)
y1 = ma.masked_invalid(array2)
m = ma.mask_or(ma.getmask(x1), ma.getmask(y1))
k = ma.array(x1, mask=m, copy=True).compressed()
j = ma.array(y1, mask=m, copy=True) .compressed()
stats.spearmanr(k,j,nan_policy='omit')
```

the computation ended in the matter of seconds. Something similar is done in the scipy itself with the exception of calling the `.compressed()` function.

I am using Python 3.5.2, Numpy 1.11.2 and scipy 0.18.1.
Many thanks,
Jakub

**Comment by WarrenWeckesser** [comment]
I can confirm that the handling of `nan` in `scipy.stats.spearmanr` is broken.  Here are some simple examples:

Base case, no `nan`s:

```
In [42]: spearmanr([1, 2, 2.5, 2], [1, 3, 2.5, 4])
Out[42]: SpearmanrResult(correlation=0.316227766016838, pvalue=0.683772233983162)
```

The default is `nan_policy='propagate'`, so the return values of the following should be `nan`:

```
In [43]: spearmanr([1, 2, 2.5, 2, np.nan], [1, 3, 2.5, 4, np.nan])
Out[43]: SpearmanrResult(correlation=0.66688592885535025, pvalue=0.21889398131323154)
```

Using `nan_policy='omit'`, the return value of the following should be the same as the base case:

```
In [44]: spearmanr([1, 2, 2.5, 2, np.nan], [1, 3, 2.5, 4, np.nan], nan_policy='omit')
Out[44]: 
SpearmanrResult(correlation=0.34999999999999998, pvalue=masked_array(data = 0.6499999999999996,
             mask = False,
       fill_value = 1e+20)
)
```

Also, `pvalue` in the above result should not be a masked array.  The use of masked arrays in `spearmanr` is an implementation detail.

**Comment by josef-pkt** [comment]
I guess `rankdata` doesn't "propagate" nans, so we don't get automatic propagation.

@jaksmid  can you time the mstats version without nans to see whether it's the implementation differences between stats and mstats version or if it's really the nan handling.
I don't see an inherent reason why the masked version should be slow.

@WarrenWeckesser  to the last point, In the multivariate case we might have missing values in the final results if for example one column (axis=0) is all nans.
So, if the stats function shouldn't return a masked array, then the results need to be `nan`ed

(aside: I was just looking at spearmanr for another purpose, but without nans)

**Comment by jaksmid** [comment]
I dug deeper and I think the explanation for the larger times and for incorrect results reported by @WarrenWeckesser are that the NaN values are not ignored.

Lets say I call `stats.spearmanr(first_input,second_input,nan_policy='omit')`, where the first input array contains nan values.

First lines of stats.spearman are:
```
contains_nan, nan_policy = _contains_nan(a, nan_policy)

    if contains_nan and nan_policy == 'omit':
        a = ma.masked_invalid(a)
        b = ma.masked_invalid(b)
        return mstats_basic.spearmanr(a, b, axis)
```
Therefore, nan values are masked in the first array and the computation goes to `mstats_basic.spearmanr`

There the or mask is computed and applied to both arrays along the lines of ` x = ma.array(x, mask=m, copy=True)`

Now comes what I believe is the problem:
 `rankx = rankdata(x)`

This does not omit the invalid values as it should but assign the rank of zero instead.
See ipython entries:
```
In [19]: j = [1,2,3,np.nan,-1,np.nan,3,3]
In [20]: x = ma.masked_invalid(j)
In [21]: k = stats.mstats_basic.rankdata(x)
In [22]: k
Out[22]: array([ 2.,  3.,  5.,  0.,  1.,  0.,  5.,  5.])
```
The ties are computed. Zeroes are counted as ties.
```
In [23]: stats.mstats_basic.count_tied_groups(k)
Out[23]: {2: 1, 3: 1}
```
This will produce the incorrect results as well as can complicate the computation (you cant use the simpler formula without ties etc).

Thoughts?

**Comment by unode** [comment]
Any progress on this? I just ran into the inconsistency with `nan`s and the `masked_array` object for pvalue.

**Comment by vickyclose** [comment]
I've been looking into the inconsistent behaviour of the spearmanr function in the case where the input variables contain nan's and repeated values (i.e. ties). I agree that the bug is due to incorrect handling of ties and I think there's a pretty simple fix - see pull request.

**Comment by Sosippus** [comment]
Any chance @WarrenWeckesser 's issue was solved? I'm finding the same problem in scipy 1.1.0, python 3.6.0. If this was fixed, could you please let me know how to solve it? I have tried updating scipy, numpy, but the issue remains. All help is welcomed!

(I have also posted this in [here](https://github.com/scipy/scipy/issues/6530#issuecomment-409238868))

**Comment by lukauskas** [comment]
So this issue is closed, but as of 1.6.1 the `spearmanr` is still quite slow when `nan_policy` is set to `omit`. 26.9s on my machine to compute rank coefficient with omitting nans vs 158ms when the lists were prefiltered. 

The answers are almost the same (within the precision margin, I would believe).

![image](https://user-images.githubusercontent.com/108413/110927324-c8c03380-8325-11eb-8fdb-b91ce0d27f66.png)

**Comment by WarrenWeckesser** [comment]
26.9 seconds is surprising.  I haven't been able to reproduce that timing.  @lukauskas, can you include a complete, self-contained example that we can run?  I just tried a similar calculation, with `x` and `y` having length 500000, each with 3000 nan's placed randomly in the arrays.  I get this:
```
In [162]: %timeit spearmanr(x, y, nan_policy='omit')                                                                                                 
664 ms ± 51.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by lukauskas** [comment]
I think there's something unique in my data that triggers the runtimes. Like you, I could not reproduce with a bunch of random numbers.

However I was able to reproduce this with shuffled version of my dataset on both a linux machine (original screenshot) and my macbook.

See [gist](https://gist.github.com/lukauskas/28c15d04875579291b6fa52b458f713a).

**Comment by mdhaber** [comment]
Hi @lukauskas. It looks like everything has been removed from the files. Could you post a regular python script?

**Comment by lukauskas** [comment]
I don't know what happened there, sorry.

I have re-uploaded the [gist](https://gist.github.com/lukauskas/28c15d04875579291b6fa52b458f713a). You will find the `data.tsv.gz` (10.6MB) there which is the most important piece of the puzzle. The script itself is simple:

```
import pandas as pd
import numpy as np
from scipy.stats import spearmanr
from sinfo import sinfo

df = pd.read_csv('data.tsv.gz', sep='\t', index_col=0)
```

This takes 30+ seconds
```
spearmanr(df['x'], df['y'], nan_policy="omit") # 30+s
```

This takes ~170ms:
```
df_no_na = df.dropna()
spearmanr(df_no_na['x'], df_no_na['y'], nan_policy="raise")
```

Versions:

```

-----
numpy       1.20.1
pandas      1.2.3
scipy       1.6.1
sinfo       0.3.1
-----
IPython             7.21.0
jupyter_client      6.1.11
jupyter_core        4.7.1
notebook            6.2.0
-----
Python 3.9.2 (default, Feb 24 2021, 13:30:36) [Clang 12.0.0 (clang-1200.0.32.29)]
macOS-10.15.7-x86_64-i386-64bit
8 logical CPU cores, i386
-----
Session information updated at 2021-03-12 19:32
```

**Comment by mdhaber** [comment]
Thanks @lukauskas. I was able to reproduce the issue with your data, too, but I still don't know why it's different from randomly generated data. gh-13312 would fix this in your case because it avoids the use of masked arrays. I tested the decorated `spearmanr` on your data, and it runs just as fast with `nan_policy='omit'` as when the `nans` are removed before the data is sent in. 

Unfortunately, `spearmanr`'s signature is unusual in that it can produce a correlation matrix, so the decorator developed in gh-13312 can't be applied to `spearmanr` without changing the way it works in some cases. Actually, in the correlation matrix case, we can't simply remove all the nans then call the routine, so I think we need to fix something on the inside.

The thing that is really slow seems to be [this loop](https://github.com/scipy/scipy/blob/5ab7426247900db9de856e790b8bea1bd71aec49/scipy/stats/mstats_basic.py#L267) in `mstats.rankdata`. The easiest patch for this might be to re-think this part.
